# Reference- and alignment-free estimation of genetic distance using k-mers

This repository contains the slides, tutorial material and (meta)data for the CBA TEA series talk given by Kevin Murray on 2017-11-17.

(If you're reading this while this sentence is here, I'm still in the process of writing the workshop. Not all material will be online.)

## Abstract

Next-gen sequencing has revolutionised the way we investigate (phylo)genetic relationships. Using traditional techniques, much effort and many resources are required to take NGS data and estimate the relationships between samples. Recent computational advances have allowed researchers to estimate these relationships directly from NGS data. In this tutorial, Kevin Murray will present a brief survey of these methods, before an interactive tutorial on their use. Participants may bring their own NGS data, or we will analyse an example dataset.


## Suggested background reading

- [Murray *et al.* 2017, **kWIP: The k-mer Weighted Inner Product**](https://doi.org/10.1371/journal.pcbi.1005727)
- [Ondov *et al.* 2016, **Mash: fast genome and metagenome distance estimation using MinHash**](https://doi.org/10.1186/s13059-016-0997-x)
